Added Js Events for Inline panel added/removed

faishal882 opened the pull request:
Fixes #9105  
## What was the issue ?
-> There were no way to capture event if a new InlinePanel is added or removed 
## Solution 
-> Added Event `wagtail:inline-panel-added` and `wagtail:inline-panel-removed` on form new Inline Panel Add/Removed







_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    -   [ ] **Please list the exact browser and operating system versions you tested**:
    -   [ ] **Please list which assistive technologies [^3] you tested**:
-   [ ] For new features: Has the documentation been updated accordingly?

**Please describe additional details for testing this change**.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

faishal882 commented:
Hey @lb-  i have made pull request but before proceeding further i would like to know if i am in correct direction of handling this issue.
And also InlinePanel delete method is written within `initChildControls` method in `InlinePanel` class i would like to separate it create its own method i think this will make more sense.
And also regarding `change` event, so this event will fire when there is change in input child node in `InlinePanel form` am i correct

Thanks in advance

lb- reviewed (changes requested):
Thanks @faishal882 for getting this started.

With the dispatched events,  we should be more explicit about the element we are dispatching against (not just finding the first not-deleted one), we should also ensure that we dispatch after any opts.callbacks are run.

## Recommendation

I think we should use the event names  `w-formset:added` & `w-formset:removed`. While this code is in InlinePanel, it's based on the core code in `client/src/components/ExpandingFormset/index.js`. We will likely migrate this code to Stimulus at some point and it will probably make sense to use `w-formset` as the core identifier for this shared logic.

This also aligns nicely with the Django naming for very similar functionality. My hope is that we can keep this event name as is when we do the change to Stimulus in the future, also `w-inline-panel` is a bit long and can confuse with the other work being done on Panels which is about collapsing/expanding panels.


### 1. added event

Let's move this right to the bottom of the `addForm` method, and add a bit of a defensive approach (checking we have a DOM element before dispatching). Adjust the name a bit also.

```js

    if (!('runCallbacks' in opts) || opts.runCallbacks) {
      if (this.opts.onAdd) this.opts.onAdd(formIndex);
      if (this.opts.onInit) this.opts.onInit(formIndex);
    }

    this.initialFocus($(`#inline_child_${newChildPrefix}-panel-content`));

    // dispatch event for initialising a form
    const newChild = $(`#inline_child_${newChildPrefix}`).get(0);
    if (!newChild) return;
    newChild.dispatchEvent(
      new CustomEvent('w-formset:added', {
        bubbles: true,
        cancelable: false,
        detail: { formIndex, ...this.opts },
      }),
    );
  }
}

```


### 2. Deletion

For the deletion, let's move this to after the control states have been updated (after the animation), adjust the name slightly. Again we are wanting to dispatch on the single DOM element and avoid checking for deleted children with the selector as this could get out of sync (especially for nested inline panels) if we are not careful.

```js
      $('#' + deleteInputId).val('1');
      currentChild.addClass('deleted').slideUp(() => {
        this.updateControlStates();
        currentChild.get(0).dispatchEvent(
          new CustomEvent('w-formset:removed', {
            bubbles: true,
            cancelable: false,
            detail: { ...this.opts },
          }),
        );
      });
    });
```

### 3. Change events

We can now patch in the change events based on when inputs are updated.

When deleted
```
      // $('#' + deleteInputId).val('1'); // replace with ....
      $('#' + deleteInputId)
        .val('1')
        .get(0)
        .dispatchEvent(new Event('change', { bubbles: true }));
```

When re-ordering
```js
    if (this.opts.canOrder) {
      /* ORDER values are 1-based, so need to add 1 to formIndex */
      $('#id_' + newChildPrefix + '-ORDER').val(formIndex + 1);
    }
```

Replace with 

```js
// ...
        .val(formIndex + 1)
        .get(0)
        .dispatchEvent(new Event('change', { bubbles: true }));
```

### 4. Testing

Add a unit test in client/src/components/InlinePanel/index.test.js to check that some of the events fire. Our JS testing is not robust here but test what you can simply.

```js
describe('InlinePanel', () => {
  # ...
  const handleAddedEvent = jest.fn();

  beforeAll(() => {
    jest.resetAllMocks();
    # ... existing HTML set up
    document.addEventListener('w-formset:added', handleAddedEvent);
  });

  it('should allow inserting a new form and calling an onAdd function', () => {
    expect(handleAddedEvent).not.toHaveBeenCalled(); // add at the start of the test
    # .. all the other test/addition
    expect(handleAddedEvent).toHaveBeenCalledWith('??'); // add at the end to check it's been called
  });
});
```

You could do something similar with writing a new test for deletion if you feel up to it, but that's not critical.

### 5. Documentation

It's important we document this, not the change events but the custom events being dispatched.

It probably makes sense here https://docs.wagtail.org/en/stable/reference/pages/panels.html#inlinepanel

We can base the docs entry on the Django docs for the similar purpose https://docs.djangoproject.com/en/4.1/ref/contrib/admin/javascript/#inline-form-events

> The InlinePanel widget will dispatch a CustomEvent when a panel is added or removed
> You may want to execute some JavaScript when an inline panel items are added or removed. The `w-formset:added` and `w-formset:removed` events allow this.

Needs fleshing out but hopefully a good starting point.

faishal882 commented:
thanks @lb-  for the input

lb- commented:
Awesome @faishal882 I'll take another look at the PR and do a bit of digging on the unit tests.

lb- commented on client/src/components/InlinePanel/index.js:
```diff
@@ -243,5 +257,16 @@ export class InlinePanel extends ExpandingFormset {
+
+    // dispatch event for initialising a form
+    const newChild = $(`#inline_child_${newChildPrefix}`).get(0);
+    if (!newChild) return;
+    newChild.dispatchEvent(
```
OK, the reason this does not work is that this guidance from me originally was wrong, sorry about that.

This specific id is only used in some cases and this cannot be made as an assumption of InlinePanel usage, instead we need to be a bit more flexible and look for the last element added to the formset. This way the template element can be one child or more children.

Here is a proposed change 

```suggestion
    this.formsElt
      .children()
      .last()
      .get(0)
      ?.dispatchEvent(
```

This will grab the formset, find the last child (newly created) and then dispatch on that element.

lb- commented on docs/reference/pages/panels.md:
````diff
@@ -109,6 +109,28 @@ Here are some built-in panel types that you can use in your panel definitions. T
+    </script>
+{% endblock %}
+```
+
````
Great start, however, I think we should try to align with a CSP compatible approach.

```
class CustomInlinePanel(InlinePanel):

    class BoundPanel(InlinePanel.BoundPanel):
        class Media:
            js = ["js/inline-panel.js"]

class MyPage(Pate):
        # .. fields

        content_panels = AbstractEmailForm.content_panels + [
            CustomInlinePanel("my_nested_fields"),
            # ... other panels
        ]
```

Then create a static JS file

```js
// static/js/inline-panel.js

document.addEventListener('w-formset:ready', function (event) {
  console.info('ready', event);
});

document.addEventListener('w-formset:added', function (event) {
  console.info('added', event);
});

document.addEventListener('w-formset:removed', function (event) {
  console.info('removed', event);
});
```

This approach means the JS will only load when you are using InlinePanel (similar to your JS version) but loads as an external script file, not an inline script.

lb- reviewed (changes requested):
@faishal882 thank you, I can see why the tests were not working and it's because I gave some wrong suggestions in my earlier review.

Sorry for the confusion, I had made the wrong assumption that all InlinePanels use that id prefix. I have added some comments for a refined approach that should allow you to write tests.

Additionally, can i ask that we also include a `'w-formset:ready'` event? I think this would add a lot of value and hopefully be a small additional change to this PR as is.

```js
export class InlinePanel extends ExpandingFormset {
  constructor(opts) {
    super(opts.formsetPrefix, opts);
    this.formsElt = $('#' + opts.formsetPrefix + '-FORMS');

    for (let i = 0; i < this.formCount; i += 1) {
      const childPrefix = this.opts.emptyChildFormPrefix.replace(
        /__prefix__/g,
        i,
      );
      this.initChildControls(childPrefix);
    }

    this.updateControlStates();
   # ... recommend we add this, delay and then dispatch a ready event
    setTimeout(() => {
      this.formsElt.get(0)?.dispatchEvent(
        new CustomEvent('w-formset:ready', {
          bubbles: true,
          cancelable: false,
          detail: { ...opts },
        }),
      );
    });
  }
```

Thank you.

faishal882 commented:
hey @lb-  added `w-formset:ready` event and also expanded tests for InlinePanel,  and also i dont think your guidance was wrong in tests in the `document.body.html` the inline child should have `id` `inline_child_emptyChildFormPrefix` but in test it was only `emptyChildFormPrefix` so when on adding form it was not able to find inline child element to dispatch event. Also updated snapshot for on adding a form and also added snapshot after deletion of form, Thanks for your input

lb- commented:
Awesome. Thanks @faishal882 I'll take another look.

lb- commented on client/src/components/InlinePanel/index.js:
```diff
@@ -243,5 +267,16 @@ export class InlinePanel extends ExpandingFormset {
     }
+
+    // dispatch event for initialising a form
+    const newChild = $(`#inline_child_${newChildPrefix}`).get(0);
+    if (!newChild) return;
+    this.formsElt.children().last().get(0)?.dispatchEvent(
```
This newChild check can be removed, it's better to just always dispatch, even if the inline child elements are not being used.

lb- commented on docs/reference/pages/panels.md:
````diff
@@ -109,6 +109,44 @@ Here are some built-in panel types that you can use in your panel definitions. T
 
 ```
 
+### JavaScript customizations in the admin  
````
```suggestion
#### JavaScript DOM events
```

Heading level should be one more deep and here is a recommended change.

lb- commented on docs/reference/pages/panels.md:
````diff
@@ -109,6 +109,44 @@ Here are some built-in panel types that you can use in your panel definitions. T
+```
+
+Then create a static JS file
+```html+django
````
```suggestion
```javascript
```

lb- commented on docs/reference/pages/panels.md:
```diff
@@ -109,6 +109,44 @@ Here are some built-in panel types that you can use in your panel definitions. T
 
+### JavaScript customizations in the admin  
+
+You may want to execute some JavaScript when an inline panel items are ready, added or removed, The ``w-formset:ready``, ``w-formset:added`` and ``w-formset:removed`` events allow this. 
```
No need for double backticks, that's only required in some of the older RST formatted docs.

lb- reviewed (approved):
This works great, thank you so much for taking on the feedback.

I have a push coming with some minor tweaks;

* Some docs typos and fixes
* Added removed event listener test
* Removed the additional snapshot tests and instead just checked for elements in the DOM, snapshots can feel like they test a lot but it's nicer to have more strict assertions about something being added/removed instead
* Removed the check for the inline panel child not existing
* Added changelog/release notes

Once pushed, I will merge when the CI all passes.
